File: .config/ringbaord/ringboard_auto_paste.py
import subprocess
import sys
import time


# ringboard-egui の実行パスを指定してください
# 環境によっては 'ringboard-egui' だけで実行できる場合もあります
RINGBOARD_COMMAND = "ringboard-egui"


def is_process_running(pid):
    """指定されたPIDのプロセスが実行中かを確認する"""
    try:
        # PID 0 は常に存在するため除外
        if pid <= 0:
            return False
        # ps コマンドでプロセス情報を取得し、指定したPIDが含まれているか確認
        # -p <PID> で特定のPIDを、-o pid= でPIDのみを出力
        result = subprocess.run(
            ["ps", "-p", str(pid), "-o", "pid="], capture_output=True, text=True
        )
        # 出力されたPIDリストに指定したPIDが含まれていれば実行中
        return str(pid) in result.stdout.strip().split()
    except Exception as e:
        print(f"プロセス実行確認中にエラーが発生しました: {e}", file=sys.stderr)
        return False


def get_ringboard_pid():
    """ringboard-egui のプロセスIDを取得する"""
    try:
        # pgrep コマンドで ringboard-egui のPIDを取得
        # -f オプションでコマンドライン全体を検索対象にする
        result = subprocess.run(
            ["pgrep", "-f", RINGBOARD_COMMAND], capture_output=True, text=True
        )
        pids = result.stdout.strip().splitlines()
        if pids:
            # 複数のPIDが見つかる可能性もあるが、ここでは最初のPIDを使用
            return int(pids[0])
        return None
    except Exception as e:
        print(f"ringboard-egui のPID取得中にエラーが発生しました: {e}", file=sys.stderr)
        return None


def simulate_ctrl_v():
    """xdotool を使って Ctrl+V をシミュレートする"""
    try:
        # xdotool がインストールされているか確認
        if (
            not subprocess.run(["which", "xdotool"], capture_output=True).returncode
            == 0
        ):
            print("エラー: xdotool がインストールされていません。", file=sys.stderr)
            print("Linuxの場合、以下のコマンドでインストールできます:", file=sys.stderr)
            print(
                "  sudo apt update && sudo apt install xdotool  (Debian/Ubuntu系)",
                file=sys.stderr,
            )
            print("  sudo dnf install xdotool  (Fedora系)", file=sys.stderr)
            print("  sudo pacman -S xdotool  (Arch Linux系)", file=sys.stderr)
            return False

        # 現在アクティブなウィンドウに対して Ctrl+V を送信
        subprocess.run(["xdotool", "key", "control+v"])
        print("Ctrl+V をシミュレートしました。")
        return True
    except Exception as e:
        print(f"Ctrl+V シミュレーション中にエラーが発生しました: {e}", file=sys.stderr)
        return False


if __name__ == "__main__":
    # ringboard-egui を起動し、終了後に一度だけペーストを実行
    print(f"{RINGBOARD_COMMAND} を起動します...")
    # result = subprocess.run([RINGBOARD_COMMAND])
    result = subprocess.run(
        ["cd /home/senken/senkenn/clipboard-history/egui && cargo run --release"],
        shell=True,
    )
    print(f"{RINGBOARD_COMMAND} が終了しました。exit code={result.returncode}")
    time.sleep(0.1)
    if result.returncode == 0:
        simulate_ctrl_v()
    else:
        print(
            "exit code が 0 ではないため、ペーストはスキップします。", file=sys.stderr
        )
    sys.exit(result.returncode)

File: .config/ringbaord/ringboard_auto_paste.rs
use std::io::{self, Write};
use std::process::{exit, Command};
use std::thread::sleep;
use std::time::Duration;

// ringboard-egui の実行コマンドを指定してください
const RINGBOARD_COMMAND: &str = "ringboard-egui";
// ringboard のソースディレクトリ（cargo run 用）
const RINGBOARD_DIR: &str = "/home/senken/senkenn/clipboard-history/egui";

fn is_process_running(pid: i32) -> bool {
    if pid <= 0 {
        return false;
    }
    match Command::new("ps")
        .args(&["-p", &pid.to_string(), "-o", "pid="])
        .output()
    {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            stdout.split_whitespace().any(|p| p == pid.to_string())
        }
        Err(e) => {
            eprintln!("プロセス実行確認中にエラーが発生しました: {}", e);
            false
        }
    }
}

fn get_ringboard_pid() -> Option<i32> {
    match Command::new("pgrep")
        .args(&["-f", RINGBOARD_COMMAND])
        .output()
    {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            stdout.lines().next().and_then(|l| l.trim().parse().ok())
        }
        Err(e) => {
            eprintln!("ringboard-egui のPID取得中にエラーが発生しました: {}", e);
            None
        }
    }
}

fn simulate_ctrl_v() -> bool {
    // xdotool がインストールされているか確認
    if !Command::new("which")
        .arg("xdotool")
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
    {
        eprintln!("エラー: xdotool がインストールされていません。");
        eprintln!("Linuxの場合、以下のコマンドでインストールできます:");
        eprintln!("  sudo apt update && sudo apt install xdotool  (Debian/Ubuntu系)");
        eprintln!("  sudo dnf install xdotool  (Fedora系)");
        eprintln!("  sudo pacman -S xdotool  (Arch Linux系)");
        return false;
    }
    match Command::new("xdotool").args(&["key", "control+v"]).status() {
        Ok(status) if status.success() => {
            println!("Ctrl+V をシミュレートしました。");
            true
        }
        Ok(status) => {
            eprintln!(
                "xdotool の実行中にエラーが発生しました。ステータス: {:?}",
                status
            );
            false
        }
        Err(e) => {
            eprintln!("Ctrl+V シミュレーション中にエラーが発生しました: {}", e);
            false
        }
    }
}

fn main() {
    println!("{} を起動します...", RINGBOARD_COMMAND);
    let status = Command::new("cargo")
        .args(&["run", "--release"])
        .current_dir(RINGBOARD_DIR)
        .status();

    match status {
        Ok(s) => {
            println!(
                "{} が終了しました。exit code={}",
                RINGBOARD_COMMAND,
                s.code().unwrap_or(-1)
            );
            sleep(Duration::from_millis(100));
            if s.success() {
                simulate_ctrl_v();
                exit(0);
            } else {
                eprintln!("exit code が 0 ではないため、ペーストはスキップします。");
                exit(s.code().unwrap_or(1));
            }
        }
        Err(e) => {
            eprintln!(
                "{} の実行中にエラーが発生しました: {}",
                RINGBOARD_COMMAND, e
            );
            exit(1);
        }
    }
}
